activerecord/lib/active_record/associations/collection_association.rb

```diff
@@ -249,7 +249,7 @@ def replace(other_array)
 
       def include?(record)
         if record.is_a?(reflection.klass)
-          if record.new_record?
+          if record.new_record? || owner.new_record?
```

Comment:
This change is needed to make [this assertion](https://github.com/rails/rails/blob/c6b227b95b631ead7ffd5c8984b16b0fbe5b3ca4/activerecord/test/cases/associations/join_model_test.rb#L494) pass, and I think it's a more correct solution too.